### PR TITLE
Adding loader for IRMAS Dataset

### PR DESCRIPTION
**Pull request by genisplaja**
### Adding loader for IRMAS Dataset

This is the new PR for the IRMAS Dataset loader, which was already commented and reviewed in [this closed PR](https://github.com/mir-dataset-loaders/mirdata/pull/288).

### Description

Please include the following information at the top level docstring for the dataset's module mydataset.py:

- [x] Describe annotations included in the dataset
- [x] Indicate the size of the datasets (e.g. number files and duration, hours)
- [x] Mention the origin of the dataset (e.g. creator, institution)
- [x] Describe the type of music included in the dataset
- [x] Indicate any relevant papers related to the dataset
- [x] Include a description about how the data can be accessed and the license it uses (if applicable)

#### Dataset loaders checklist:

- [x] Create a script in `scripts/`, e.g. `make_my_dataset_index.py`, which generates an index file.
- [x] Run the script on the canonical version of the dataset and save the index in `mirdata/indexes/` e.g. `my_dataset_index.json`.
- [x] Create a module in mirdata, e.g. `mirdata/my_dataset.py`
- [x] Create tests for your loader in `tests/`, e.g. `test_my_dataset.py`
- [x] Add your module to `docs/source/mirdata.rst` and `docs/source/datasets.rst`
- [x] Add the module to `mirdata/__init__.py`
- [x] Add the module to the list in the `README.md` file, section `Currently supported datasets`
- [x] Docs are building
- [x] test_full_dataset is passing correctly

#### Please-do-not-edit flag
To reduce friction, we will make commits on top of contributor's pull requests by default unless they use the `please-do-not-edit` flag. If you don't want this to happen don't forget to add the flag when you start your pull request.

### CODE REVIEW

**Comment by nkundiushuti** [comment]
this would need to be updated with the recent PRs in master

**Comment by genisplaja** [comment]
> this would need to be updated with the recent PRs in master

Yes, thanks! Done now!

**Review by nkundiushuti — APPROVED** [review]

File: mirdata/datasets/irmas.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+  doi          = {10.5281/zenodo.1290750},
+  url          = {https://doi.org/10.5281/zenodo.1290750}
+"""
+
+REMOTES = {
+    'training_data': download_utils.RemoteFileMetadata(
+        filename='IRMAS-TrainingData.zip',
+        url='https://zenodo.org/record/1290750/files/IRMAS-TrainingData.zip?download=1',
+        checksum='4fd9f5ed5a18d8e2687e6360b5f60afe',  # the md5 checksum
+        destination_dir=None,  # relative path for where to unzip the data, or None
```
Comment (rabitt) [inline mirdata/datasets/irmas.py:99]: nit: remove the comment here and below

File: mirdata/datasets/irmas.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    ),
+}
+
+
+DATA = utils.LargeData(
+    'irmas_index.json'
+)
+
+
+INST_DICT = {
```
Comment (rabitt) [inline mirdata/datasets/irmas.py:127]: Should this be a list instead of a dict?

File: mirdata/datasets/irmas.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    """IRMAS track class
+
+    Args:
+        track_id (str): track id of the track
+        data_home (str): Local path where the dataset is stored.
+            If `None`, looks for the data in the default directory, `~/mir_datasets/Mridangam-Stroke`
+
+    Attributes:
+        track_id (str): track id
+        predominant_instrument (list): Training tracks predominant instrument
```
Comment (rabitt) [inline mirdata/datasets/irmas.py:142]: is the description of the different instrument categories documented somewhere?

File: mirdata/datasets/irmas.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        self.drum = None
+        self.train = True
+
+        # TRAINING TRACKS
+        if '__' in track_id:
+            self.predominant_instrument = os.path.basename(os.path.dirname(self.audio_path))
+            assert self.predominant_instrument in INST_DICT, "Instrument {} not in instrument dict".format(self.predominant_instrument)
+
+            # Drum presence annotation is present
+            if 'dru' in self._track_paths['audio'][0] or 'nod' in self._track_paths['audio'][0]:
```
Comment (rabitt) [inline mirdata/datasets/irmas.py:173]: nit: in __init__, since you use `self._track_paths['audio'][0]` a lot here, maybe create:
`self._audio_filename = self._track_paths['audio'][0]`, and use it everywhere else.

File: mirdata/datasets/irmas.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+    pred_inst = []
+    with open(annotation_path, 'r') as fopen:
+        pred_inst_file = fopen.readlines()
+        for inst_ in pred_inst_file:
+            inst_code = inst_[:3]
+            assert inst_code in INST_DICT, "Instrument {} not in instrument dictionary".format(inst_code)
+            pred_inst.append(inst_code)
+
+        return pred_inst
```
Comment (rabitt) [inline mirdata/datasets/irmas.py:250]: maybe I don't understand the dataset well - I'd expect "predominant instrument" to have only one option. Are these annotations possibly giving all the instruments, not just the predominant?

Asking to know if the labels for `predominant_instrument` in the training set have the same properties/definition as the ones in the test set. If they're different, maybe we could just rename this function.

cc @nkundiushuti

**Review by rabitt — COMMENTED** [review]
Very little nits in the comments. Don't forget to add the loader to the README as well.

New loader checklist:
- [x] Create a script in `scripts/`, e.g. `make_my_dataset_index.py`
- [x] Run the script on the canonical version of the dataset and save the index in `mirdata/indexes/` e.g. `my_dataset_index.json`
- [x] Create a module in `mirdata`, e.g. `mirdata/my_dataset.py`
- [x] Create tests in `tests/`, e.g. `test_my_dataset.py`
- [x] Add the module to `docs/source/mirdata.rst`
- [x] Add the module to `docs/source/datasets.rst`
- [x] Add the module to `mirdata/__init__.py`
- [x] Add the module to the list in the README.md file, section Currently supported datasets
- [ ] Run test_full_dataset.py
- [x] Make sure the docs build properly

File: mirdata/datasets/irmas.py (hunk 5 above)
Comment (genisplaja) [reply mirdata/datasets/irmas.py:250]: The instrument annotations contain just the predominant instrument for both training and testing. There are no annotations about the whole instrument line-up of the tracks. 
However, the training set annotation contain just a single predominant instrument, while the testing set annotations contain more that one predominant instrument, if applicable. For instance, in a piece where the guitar and the voice are similarly predominant, the testing set annotation will contain both instruments.

File: mirdata/datasets/irmas.py (hunk 3 above)
Comment (genisplaja) [reply mirdata/datasets/irmas.py:142]: As stated in the comment below, there is just a single instrument category: the predominant instrument. There is no information about the whole instrument line-up of the track, just the predominant one. The important point is that, for the training set there is just a single predominant instrument annotated, while for the testing set there can be more than one predominant instrument in the annotation.
However I was trying to use `instrument` as a catched property to mantain the terminology with the rest of the datasets.

Comment (rabitt) [reply mirdata/datasets/irmas.py:142]: Ah i mean the definitions of the three letter instrument codes/ how each instrument category is defined.

**Review by rabitt — APPROVED** [review]
👍

File: mirdata/datasets/irmas.py (hunk 3 above)
Comment (genisplaja) [reply mirdata/datasets/irmas.py:142]: It's in the attached link, but I will add it to the loader `irmas.py` just in case!

Comment (genisplaja) [reply mirdata/datasets/irmas.py:142]: Ah yes! It's explained in the documentation on the top of the loader file `irmas.py`. It's also stated in the links attached to the documentation.

**Comment by genisplaja** [comment]
Thanks for the reviews @nkundiushuti, @rabitt and @magdalenafuentes! I just added an extra assert line to check that the genre codes are correctly parsed. Should we merge this PR then? Cheers!

**Comment by nkundiushuti** [comment]
> Thanks for the reviews @nkundiushuti, @rabitt and @magdalenafuentes! I just added an extra assert line to check that the genre codes are correctly parsed. Should we merge this PR then? Cheers!

yes. feel free to merge
